source/common/conn_pool/conn_pool_base.h

```diff
@@ -102,7 +114,10 @@ class ActiveClient : public LinkedObject<ActiveClient>,
   // Sets the remaining streams to 0, and updates pool and cluster capacity.
   virtual void drain();
 
-  virtual bool hasHandshakeCompleted() const { return state_ != State::Connecting; }
+  virtual bool hasHandshakeCompleted() const {
+    ASSERT(!supportsEarlyData());
```

Comment:
Presumably, clients which support early data need to override this method?

Reply:
Correct.